### PR TITLE
New JobGuard application code

### LINKED ISSUE #3454: JobGuard should fetch required status checks from GitHub.
**Description**
WIth the current implementation of JobGuard it checks the local YAMLs for the required status checks. If the job has been removed already from the master branch, but its definition is still present in the release branch then JobGuard will try to wait for the job.

In the new workflow if the job is not meant to be run it will never be reported to GitHub. Job guard will also wait infinitely for the job to be reported.

**Reasons**
This rewrite is mandatory before introducing Tide.

**Acceptance Criteria**
- JobGuard picks active status checks from the GitHub API, not from local YAML definitions.
- JobGuard fails if any of the required checks have failed (?)
- JobGuard should have the ability to block the pipeline execution until the previous checks have passed
- Maybe we should leverage JobGuard as separate init container? It will be much easier to maintain and bump if necessary

### PR DESCRIPTION
**Pull request by Ressetkk**
**Description**
Rewritten JobGuard to use Prow's GitHub client for fetching, building and updating the statuses. It won't depend on the local prowjob definition yamls, but will build the required contexts based on initial ones that are present on the commit. Then JobGuard will periodically update and check the statuses.

**Related issue(s)**

Closes #3454

### CODE REVIEW

**Review by adamwalach — APPROVED** [review]

**Review by adamwalach — DISMISSED** [review]

File: development/jobguard/README.md [hunk 1]
```diff
@@ -11,24 +11,30 @@ The main purpose of Job Guard is to delay running integration jobs that depend o
...
+|Flag|Required|Description|
+|---|---|---|
+|`-github-host`| No | GitHub's default host.|
+|`-github-endpoint`| No | GitHub API endpoint.|
+|`-github-graphql-endpoint`| No | GitHub GraphQL API endpoint.|
+|`-github-token-path`|Yes|Path to the file containing the GitHub OAuth secret.|
+|`-debug`|No|Enable debug logging.|
+|`-dry-run`|No|Run in dry mode.|
+|`-expected-contexts-regexp`|Yes|Regular expression with expected contexts.|
+|`-fail-on-no-contexts`|No|Fail if regexp does not match to any of the GitHub contexts.|
```
Comment (majakurcius) [inline development/jobguard/README.md:23]:
```suggestion
|`-fail-on-no-contexts`|No|Fail if regexp does not match any of the GitHub contexts.|
```
Use `match` in this way: `A matches B` or `You match A to B` or `A is a match for B`.

File: development/jobguard/README.md [hunk 2]
```diff
@@ -11,24 +11,30 @@ The main purpose of Job Guard is to delay running integration jobs that depend o
...
+|`-timeout`|No|Time after the JobGuard fails.|
```
Comment (majakurcius) [inline development/jobguard/README.md:24]:
```suggestion
|`-timeout`|No|Time after JobGuard fails.|
```
You don't use `the` anywhere else.

File: development/jobguard/README.md [hunk 3]
```diff
@@ -11,24 +11,30 @@ The main purpose of Job Guard is to delay running integration jobs that depend o
 
 To run the application, use this command:
```
Comment (majakurcius) [inline development/jobguard/README.md:13]: After removing the command you forgot about this line. 
Please move the command from the **Example** section here, and remove the `### Run the application` line above. 
See also the READMEs for [prowjobparser](https://github.com/kyma-project/test-infra/tree/cef5cedd57404c2aa971662f9bff926ad3b90285/development/prowjobparser) and [image-syncer](https://github.com/kyma-project/test-infra/tree/cef5cedd57404c2aa971662f9bff926ad3b90285/development/image-syncer). 
``````suggestion
To run the application, use this command:

```shell
go run cmd/jobguard/main.go \
  -github-endpoint=http://ghproxy \
  -github-endpoint=https://api.github.com \
  -github-token-path=/path/to/oauth \
  -org=example-org \
  -repo=example-repo \
  -base-ref=13abc \
  -expected-contexts-regexp="(some-context-regexp|another-context)"
```
``````

File: development/jobguard/README.md [hunk 4]
```diff
@@ -11,24 +11,30 @@ The main purpose of Job Guard is to delay running integration jobs that depend o
...
 
-```bash
-COMMIT_SHA={commit_sha} PROW_CONFIG_FILE={prow_config_file} PROW_JOBS_DIRECTORY={prow_jobs_directory} go run cmd/main.go
+|Flag|Required|Description|
```
Comment (majakurcius) [inline development/jobguard/README.md:14]: Please add this section name here. I'd go with `Flags`, but [prowjobparser](https://github.com/kyma-project/test-infra/tree/cef5cedd57404c2aa971662f9bff926ad3b90285/development/prowjobparser) already uses `CLI parameters`. 

```suggestion
### CLI parameters

|Flag|Required|Description|
```

File: development/jobguard/README.md [hunk 5]
```diff
@@ -11,24 +11,30 @@ The main purpose of Job Guard is to delay running integration jobs that depend o
...
+```shell
+go run cmd/jobguard/main.go \
+  -github-endpoint=http://ghproxy \
+  -github-endpoint=https://api.github.com \
+  -github-token-path=/path/to/oauth \
+  -org=example-org \
+  -repo=example-repo \
+  -base-ref=13abc \
+  -expected-contexts-regexp="(some-context-regexp|another-context)"
 ```
```
Comment (majakurcius) [inline development/jobguard/README.md:40]:
```suggestion
```

Please move this to the top of the **Usage** section. See my comments above.

File: development/jobguard/README.md [hunk 6]
```diff
@@ -7,28 +7,32 @@ The main purpose of Job Guard is to delay running integration jobs that depend o
...
-| **API_ORIGIN**            | No        | `https://api.github.com`  | Origin of the GitHub API |
-| **REPO_OWNER**            | No        | `kyma-project`            | Username or organization name of the repository owner |
-| **REPO_NAME**             | No        | `kyma`                    | Name of the repository |
-| **COMMIT_SHA**            | Yes       | None                          | Commit SHA |
-| **GITHUB_TOKEN**          | No        | None                          | Authorization token for GitHub API |
-| **JOB_NAME_PATTERN**      | No        | `components`              | Regexp to filter dependant statuses |
-| **PROW_CONFIG_FILE**      | Yes       | None                          | Path to the Prow `config.yaml` file  |
-| **PROW_JOBS_DIRECTORY**   | Yes       | None                          | Path to the directory with Prow jobs |
+## CLI parameters
+
```
Comment (majakurcius) [inline development/jobguard/README.md:23]:
```suggestion

The JobGuard accepts the following command line parameters:

```
`Job Guard` or `JobGuard`? You use one in the description and the other in the table with flags. Please align.

**Review by majakurcius — APPROVED** [review]

**Review by klaudiagrz — APPROVED** [review]

**Review by kazydek — APPROVED** [review]

**Review by adamwalach — APPROVED** [review]

**Pull request merged**